iommi/declarative/namespace.py

```diff
@@ -62,18 +74,20 @@ def setitem_path(self, path, value):
                 self[key] = value
 
     def __repr__(self):
+        # NB: repr is called on any values in the namespace
         return "%s(%s)" % (type(self).__name__, ", ".join('%s=%r' % (k, v) for k, v in sorted(flatten_items(self), key=lambda x: x[0])))
 
     def __str__(self):
+        # NB: str is called on any values in the namespace
         return "%s(%s)" % (type(self).__name__, ", ".join('%s=%s' % (k, v) for k, v in sorted(flatten_items(self), key=lambda x: x[0])))
 
     def __call__(self, *args, **kwargs):
         params = Namespace(self, kwargs)
 
         try:
             call_target = params.pop('call_target')
-        except KeyError:
-            raise TypeError('Namespace was used as a function, but no call_target was specified. The namespace is: %s' % self)
+        except KeyError as e:
+            raise TypeError('Namespace was used as a function, but no call_target was specified. The namespace is: %s' % self) from e
```

Comment:
Why the `from e`? What does that add?

Reply:
I'll admit that I also wasn't aware of what this means until Sourcery [suggested](https://docs.sourcery.ai/suggestions/raise-from-previous-error/) it to me. Essentially, it makes Python aware that the `TypeError` we raise here is _directly_ caused by the previous `KeyError`, rather than an unrelated exception that happened to occur while the `KeyError` was being handled.

Here is a nice write-up: https://stefan.sofa-rockers.org/2020/10/28/raise-from/

Incidentally, TIL from this post that if one ever wants to suppress the underlying cause to get a clean traceback, one can write: `raise ... from None`. I'll probably start using this frequently in my own code now. ;-P